### PR TITLE
Add Ubisoft connect (Linux Proton) support

### PR DESCRIPTION
**Pull request by Ofacy**
This code might not look good, but it works on my machine hey.
I just did the bare minimum to get it working on my steam deck, just wanted to share this.

Quick note about that commit, this is my first time reading and writing rust, this code might not look good or even work on your machine, but I managed to get it working on my steam deck and I wanted to share this with anyone who wants it, so here is my code.

### CODE REVIEW

**Comment by Ofacy** [comment]
This would resolve #255 and #215.

**Comment by PhilipK** [comment]
This is seriously awesome, thanks! I will look through the code in detail tonight, but great they you got it working

**Comment by PhilipK** [comment]
For being the first time you have done any Rust I am very impressed.

The errors you are seeing in the pipeline is because I have configured clippy (linter) to disallow any operation that would possibly crash Boilr. I have made the changes locally that would fix this, I guess the best way of getting them into this pull request is by using the suggestions feature.

File: src/platforms/uplay/platform.rs [hunk 1]
```diff
@@ -16,19 +21,92 @@ pub struct UplayPlatform {
...
+                    res.exe_path = uplay_exe_path;
+                    res.games_path = uplay_games;
+                    res.compat_folder = Some(dir.path());
+                    return Ok(res);
+                }
+            }
+        }
+    }
+    return Err(eyre::eyre!(
+        "Could not find uplay launcher"))
```
Comment (PhilipK) [inline src/platforms/uplay/platform.rs:103]:
```suggestion
   Err(eyre::eyre!(
        "Could not find uplay launcher"))
```

File: src/platforms/uplay/platform.rs [hunk 2]
```diff
@@ -90,6 +171,104 @@ fn get_games_from_winreg() -> eyre::Result<Vec<UplayGame>> {
...
+                    launcher: launcher_path.exe_path.clone(),
+                    launcher_compat_folder: launcher_path.compat_folder.clone(),
+                    launch_id,
+                });
+                launch_id = launch_id + 1;
+            }
+        }
+    }
+    Ok(games)
+}
```
Comment (PhilipK) [inline src/platforms/uplay/platform.rs:270]:
```suggestion
#[cfg(target_family = "unix")]
fn get_games_from_proton() -> eyre::Result<Vec<UplayGame>> {
    let mut games = vec![];

    let launcher_path = get_launcher_path()?;
    let parent = launcher_path.exe_path.parent().unwrap_or_else(|| Path::new("/"));
    let file = File::open(parent
    .join("cache")
    .join("configuration")
    .join("configurations"))?;
    let mut reader = BufReader::new(file);
    let mut buffer = Vec::new();
    
    // Read file into vector.
    reader.read_to_end(&mut buffer)?;

    let mut splits: Vec<String> = Vec::new();

    while !buffer.is_empty() {
        
        let foundindex: usize = match find_subsequence(buffer.as_slice(), b"version: 2.0") {
            Some(index) => {index},
            None => {break;},
        };
        let (mut first, second) = buffer.split_at(foundindex);
        if first.len() >= 14usize {
            first = first.split_at(first.len()-14).0;
        }
        splits.push(unsafe {std::str::from_utf8_unchecked(first).to_string()});
        buffer = second.split_at("version: 2.0".len()).1.to_vec();
    }

    
    

    for gameconfig in splits {
        if !gameconfig.contains("executables:") {continue};
        if !gameconfig.contains("online:") {continue};
        if !gameconfig.contains("shortcut_name:") {continue};
        if !gameconfig.contains("register:") {continue};

        let mut inonline = false;
        let mut shortcut_name: String = "".to_string();
        let mut game_id: String = "".to_string();
        let mut icon_image: PathBuf = "".into();
        let mut launch_id = 0;
        for line in gameconfig.split('\n') {
            let trimed = line.trim();
            if trimed.starts_with("online:") {
                inonline = true;
                continue;
            }
            if trimed.starts_with("offline:") {
                break;
            }
            if trimed.starts_with("icon_image: ") {
                let split = trimed.split_at("icon_image: ".len()).1;
                if split.is_empty()  {break}; // invalid config.
                icon_image = parent.join("data").join("games").join(split);
            }
            if !inonline {continue};
            if trimed.starts_with("- shortcut_name:") {
                let split = trimed.split_at("- shortcut_name:".len()).1;
                if split.is_empty()  {break}; // invalid config.
                shortcut_name = split.to_string();
                continue;
            }

            if trimed.starts_with("register: ") {
                let split = trimed.split_at("register: ".len()).1;
                if split.is_empty()  {break}; // invalid config.
                game_id = split
                .strip_prefix("HKEY_LOCAL_MACHINE\\SOFTWARE\\Ubisoft\\Launcher\\Installs\\").unwrap_or_default()
                .strip_suffix("\\InstallDir").unwrap_or_default().to_string();
                continue;
            }

            if trimed == "denuvo: yes" {
                games.push(UplayGame {
                    name: shortcut_name.clone(),
                    icon: icon_image.to_string_lossy().to_string(),
                    id: game_id.clone(),
                    launcher: launcher_path.exe_path.clone(),
                    launcher_compat_folder: launcher_path.compat_folder.clone(),
                    launch_id,
                });
                launch_id += 1;
            }
        }
    }
    Ok(games)
}
```

**Review by PhilipK — COMMENTED** [review]
Here suggestions that should fix the possible crashes

**Comment by Ofacy** [comment]
There it should now successfully build for linux too,  hope this works.

**Comment by PhilipK** [comment]
The last problem for window just seems to be a function that needs to be included in that built. Do you have access to a windows machine? Otherwise i can find the fix tonight

**Comment by Ofacy** [comment]
There this should fix the error, didn't have much time, we might want to make those includes "unix only" to remove the last warnings.

**Pull request merged**